content/blog/2025-05-21-performance-optimization-methodology-for-valkey/index.md

```diff
@@ -0,0 +1,159 @@
+
+While parallelism allows us to "use more resources efficiently," efficiency optimizations enable us to "do more with the same resources." Both approaches are essential in a comprehensive optimization strategy, particularly for mission-critical systems like Valkey.
+
+This article focuses primarily on the second strategy, exploring methodical approaches to improving CPU efficiency with concrete examples from our contributions to the Valkey codebase.
+
+
+## Benchmarking: The Foundation for Reliable Performance Optimization
+
+Reliable performance optimization requires consistent, reproducible measurements to evaluate whether code changes actually improve performance. Without methodical benchmarking in a controlled environment, it's impossible to accurately quantify improvements or determine if optimizations introduce regressions in other areas.
+
+To isolate CPU performance factors and eliminate variables that might affect measurements, we've implemented the following constraints:
+
+1. **Bare-metal servers** rather than virtual machines to eliminate hypervisor overhead and contention
+2. **Process pinning** using taskset to pin Valkey to specific cores, preventing thread migration overhead
+3. **Local network interfaces** (loopback) for client-server communication to minimize network variability
+4. **High CPU utilization** benchmark parameters to ensure we're measuring true CPU performance limits
+
+This controlled environment allows us to accurately attribute improvements to specific code optimizations, providing reliable measurements of throughput and latency improvements. For each optimization attempt, we establish baseline performance metrics, implement changes in isolation, and then re-measure to quantify the impact. This disciplined approach ensures that our optimizations deliver genuine benefits rather than illusory improvements that might disappear in production environments.
+
+## Optimization Methodology
+
+In the following sections, we'll share practical examples of optimization techniques applied to Valkey. These insights represent our modest contributions that might be helpful for others working on similar performance challenges.
+
+### 1. Eliminating Redundant Code
+
+Simplifying execution paths by removing redundant operations is a straightforward optimization approach, though identifying truly unnecessary code requires careful analysis.
+
+#### How to Identify Redundant Code
+
+The key to finding redundant code lies in having:
+
+1. **Representative Workloads**: Test workloads that reflect real-world usage patterns
+2. **Proper Profiling Tools**: Tools like `perf` and Intel® VTune™ Profiler to identify hot code paths
```

Comment:
Do you want to add a link to VTune - it feels relevant? If not, don't worry about this.

Reply:
added :)